plugins/inventory/digitalocean.py

```diff
@@ -0,0 +1,205 @@
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, Cacheable
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+
+    NAME = 'community.digitalocean.digitalocean'
+
+    def verify_file(self, path):
+        valid = False
+        if super(InventoryModule, self).verify_file(path):
+            if path.endswith(
+                    ('do_hosts.yaml', 'do_hosts.yml',
+                     'digitalocean.yaml', 'digitalocean.yml',
+                     'digital_ocean.yaml', 'digital_ocean.yml')
+            ):
+                valid = True
+            else:
+                self.display.vvv(
+                    'Skipping due to inventory source file name mismatch. '
+                    'The file name has to end with one of the following: '
+                    'do_hosts.yaml, do_hosts.yml '
+                    'digitalocean.yaml, digitalocean.yml, '
+                    'digital_ocean.yaml, digital_ocean.yml.')
+        return valid
+
+    def parse(self, inventory, loader, path, cache=True):
+        super(InventoryModule, self).parse(inventory, loader, path)
+
+        # cache settings
+        self._read_config_data(path)
+        self.cache_key = self.get_cache_key(path)
+
+        self.use_cache = self.get_option('cache') and cache
+        self.update_cache = self.get_option('cache') and not cache
+
+        results = []
+        if not self.update_cache:
+            try:
+                results = self._cache[self.cache_key]['digitalocean']
+            except KeyError:
+                pass
+
+        if not results:
+            if self.cache_key not in self._cache:
+                self._cache[self.cache_key] = {'digitalocean': ''}
+
+        # request parameters
+        base_url = 'https://api.digitalocean.com/v2'
+        resource = 'droplets'
+        pagination = self.get_option('pagination')
+        url = '{0}/{1}?per_page={2}'.format(base_url, resource, pagination)
+
+        api_token = self.get_option('api_token')
+        headers = {
+            'Content-Type': 'application/json',
+            'Authorization': 'Bearer {0}'.format(api_token)
+        }
+
+        # send request
+        req = Request(headers=headers)
+        try:
+            self.display.vvv('Sending request to {0}'.format(url))
+            response = json.load(req.get(url))
+            key = 'droplets'
+            payload = response.get(key)
+            if not payload:
+                raise AnsibleParserError(
+                    "There is no droplet information received")
+            if response.get('links'):
+                while response.get('links').get('pages').get('next'):
+                    url = response.get('links').get('pages').get('next')
+                    self.display.vvv('Sending request to {0}'.format(url))
+                    response = json.load(req.get(url))
+                    payload += response.get(key)
```

Comment:
This part is a bit awkward and convoluted. Maybe we should simplify it into something like this:

```suggestion
        # request parameters
        api_token = self.get_option('api_token')
        headers = {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer {0}'.format(api_token)
        }

        # send request
        req = Request(headers=headers)
        url = 'https://api.digitalocean.com/v2/droplets?per_page' + str(pagination)
        payload = []
        try:
            while url:
                self.display.vvv('Sending request to {0}'.format(url))
                response = json.load(req.get(url))
                payload.extend(response['droplets'])
                url = response.get('links', {}).get('pages', {}).get('next')
```

Reply:
wow, that's elegant. After I committed this I started to think of a 'do while' like approach, but that's certainly better. Thank you! I still don't understand the empty maps in the last line, but I believe that's good. It's above my skills.

Reply:
One thing I rolled back (by accident the pagination variable went to trash with the other garbage). The url line was too long (I don't know how strict Ansible follows the pep8 recommendations). So I decided to break the url to logical parts, because it makes easier to adopt it to other cases or move the url builder to a separate class like the DigitalOceanHelper class in module_utils. What do you think?

Reply:
If the refactoring ever happens, most of the code related to DO API interaction will be removed and replaced by a generalized HTTP API client from [https://github.com/ansible-collections/community.digitalocean/blob/main/plugins/module_utils/digital_ocean.py). So whatever we do here is good enough.

BTW, that utility module is quite good, but in the current form, it cannot be used from non-module code. With a little help, it could be transformed into something really nice.

As for the line length limit, Ansible in general standardized on 160 characters per line, and this limit is enforced by the sanity tests.

Reply:
You convinced me, I simplified the url line